Variation of `core/buttons` block for `Add to cart` button

imanish003 opened the pull request:
Variation of `core/buttons` block for Add to Cart button.

#### Why?
The “Add to Cart” button is missing some of the core buttons’ functionalities. I compared Add to cart button with Core button & identified the following differences:
| Add to cart button | Core button |
|-|-|
| ![image](https://user-images.githubusercontent.com/16707866/211290942-a985077e-02a3-4901-aa86-9e6840a8035d.png) | ![image](https://user-images.githubusercontent.com/16707866/211290991-e9a34057-3ff5-42c6-8a4a-87b736a33e1e.png) |

**Following are the extra functionality in Core Button:**
- Styles: Fill & Outline
- Width Settings
- Typography
    - Font Family
    - Appearance
    - Line height
    - Letter spacing
    - Decoration
    - Letter case

**We want all of these extra functionalities to be available in Add to cart button. Therefore, We want to extend core/buttons instead of duplicating all the available functionality.**

Fixes #8085 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1.
2.
3.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Implemented `Add to cart` button using `core/buttons` variation

imanish003 committed:
Creating variation of core/button

imanish003 committed:
Improvements using products dynamic data

imanish003 committed:
More improvements

imanish003 committed:
Replace core/button with core/buttons

imanish003 committed:
Add comment

imanish003 committed:
Pass props to child component i.e. core/button

imanish003 committed:
Fix: Text doesn't reflect everywhere in Editor

imanish003 committed:
Hide link toolbar button for core/buttons variation

imanish003 committed:
Internationalize Add to Cart button text

imanish003 committed:
Fix minor issues

imanish003 committed:
Code quality improvements

imanish003 committed:
Add comments for readability of code

imanish003 committed:
Fix: Text nodes not getting deleted

sunyatasattva commented on assets/js/blocks/product-query/constants.ts:
```diff
@@ -9,6 +9,7 @@ import type { InnerBlockTemplate } from '@wordpress/blocks';
  */
 import { QueryBlockAttributes } from './types';
 import { VARIATION_NAME as PRODUCT_TITLE_ID } from './variations/elements/product-title';
+import { VARIATION_NAME as ADD_TO_CART_BUTTON_VARIATION_NAME } from './variations/elements/add-to-cart-button';
```
Just for consistency with the other variables, I'd name this `ADD_TO_CART_BUTTON_ID`.

imanish003 replied:
Thanks for pointing this out. IMO It's better to add `VARIATION_NAME` suffix to make names more readable. For example in [this file](https://github.com/woocommerce/woocommerce-blocks/blob/0f2903ce4c5465d84d2f5f145e23d505a08f48ae/assets/js/blocks/product-query/index.tsx) also, we are using the same name `PRODUCT_TITLE_ID` but value it contains is core name.

What do you think about changing the names to add suffix `CORE_NAME` & `VARIATION_NAME` instead for all names?

sunyatasattva replied:
I don't have a strong opinion about this. If you strongly feel that the further variable name length is paid off in readability, go for it! 😊  As long as we are using the same names consistently.

imanish003 replied:
Great! Thanks. I will update the names 🤝 🙌🏻

sunyatasattva commented on assets/js/blocks/product-query/index.tsx:
```diff
@@ -11,6 +11,7 @@ import { isWpVersion } from '@woocommerce/settings';
 import { CORE_NAME as PRODUCT_TITLE_ID } from './variations/elements/product-title';
 import { CORE_NAME as PRODUCT_SUMMARY_ID } from './variations/elements/product-summary';
 import { CORE_NAME as PRODUCT_TEMPLATE_ID } from './variations/elements/product-template';
+import { CORE_NAME as ADD_TO_CART_BUTTON_CORE_NAME } from './variations/elements/add-to-cart-button';
```
Same here.

sunyatasattva commented on assets/js/blocks/product-query/variations/elements/add-to-cart-button.tsx:
```diff
@@ -0,0 +1,109 @@
+	'woo-gutenberg-products-block'
+);
+const BLOCK_ICON: JSX.Element = (
+	<Icon icon={ button } className="wc-block-editor-components-block-icon" />
```
Do we need to change the icon?

imanish003 replied:
Nice catch. I will remove it.

sunyatasattva commented on assets/js/blocks/product-query/variations/elements/add-to-cart-button.tsx:
```diff
@@ -0,0 +1,109 @@
+export const CORE_NAME = 'core/buttons';
+export const VARIATION_NAME = 'woocommerce/product-add-to-cart-button';
+
+if ( isFeaturePluginBuild() ) {
```
We decided to remove the feature gate from element variations here: https://github.com/woocommerce/woocommerce-blocks/pull/8297 so this guard is no longer required :)

sunyatasattva commented on assets/js/blocks/product-query/variations/elements/add-to-cart-button.tsx:
```diff
@@ -0,0 +1,109 @@
+}
+
+/**
+ * This helps us in modifying the UI in editor
+ * for the button block variation
```
Super nitpick: but I don't think this comment needs a line break. Eyeballing it, it seems to fit in the line length, am I wrong?

imanish003 replied:
You are right. Nice catch.

sunyatasattva commented on assets/js/blocks/product-query/variations/elements/add-to-cart-button.tsx:
```diff
@@ -0,0 +1,109 @@
+
+			if ( isWoocommerceVariation && ! props.attributes?.text?.length ) {
+				// Set initial attributes when block is added to the editor.
+				props.setAttributes( {
+					...props.attributes,
+					text: __( 'Add to cart', 'woo-gutenberg-products-block' ),
+					style: ! props.attributes.style && {
+						spacing: {
+							padding: {
+								top: 'var:preset|spacing|30',
+								right: 'var:preset|spacing|40',
+								bottom: 'var:preset|spacing|30',
+								left: 'var:preset|spacing|40',
+							},
+						},
+					},
+					align: 'center',
+					fontSize: 'small',
+				} );
```
Instead of setting them dynamically here, you can actually pass the default attributes to the `registerBlockVariation` function. Since you are using the `registerElementVariation` util, you might want to add this as an argument to that function.

imanish003 replied:
I am not sure if we can pass attributes in `registerBlockVariation` because we are registering `core/buttons` as a variation, but we have to change attributes for `core/button`, which is children of `core/buttons`. 

I also tried using `innerBlocks` as follows, but it didn't work.
```
registerElementVariation( CORE_NAME, {
	blockDescription: BLOCK_DESCRIPTION,
	blockTitle: BLOCK_TITLE,
	variationName: VARIATION_NAME,
	innerBlocks: [
		[
			'core/button',
			{
				text: __( 'Add to cart', 'woo-gutenberg-products-block' ),
				align: 'center',
			},
			[],
		],
	],
} );
```
You can see the code I tried in [this commit](https://github.com/woocommerce/woocommerce-blocks/commit/6c9069e9cb65abf11271ded9f855f905fc489c39).

Am I missing something? 🤷🏻‍♂️

sunyatasattva replied:
Mm, I don't think you are missing something, this code seems correct to me. What didn't work about passing the attributes to this `innerBlocks` configuration?

sunyatasattva commented on assets/js/blocks/product-query/variations/elements/add-to-cart-button.tsx:
```diff
@@ -0,0 +1,109 @@
+								'.edit-post-visual-editor button[name="link"]'
+							);
+						if ( linkElement ) linkElement.style.display = 'none';
+					}, 0 );
```
You might want to use `useEffect` here instead of a `setTimout` with `0` delay.

imanish003 replied:
Thanks for pointing this out. While experimenting, I could not make it work using `useEffect`. But today I tried again with clear mind and it worked([commit link](https://github.com/woocommerce/woocommerce-blocks/commit/83dc133564e535c3283c1db82d2389e7b462e224)) 🙌🏻

imanish003 replied:
While I was testing it thoroughly, I found that sometimes link toolbar button is not hiding. I am debugging it. 
Apart from this, everything should work as expected. Please let me know if you find any other issue in current implementation 🙂

imanish003 replied:
**Update:**:
It happens because sometimes `core/button` doesn't re-render, but toolbar menu re-renders. 
One such scenario is when using ESC key and clicking the RichText field, as shown below video:

https://user-images.githubusercontent.com/16707866/215081704-527a4cb0-210b-40c5-a73b-c0d81b899fc4.mov

Looking more into it.

sunyatasattva commented on assets/js/blocks/product-query/variations/elements/add-to-cart-button.tsx:
```diff
@@ -0,0 +1,109 @@
+					}, 0 );
+				}
+
+				return <BlockEdit { ...props } />;
```
This is not needed, as you are returning below anyways.

imanish003 replied:
Oh yes, This is a leftover from previous experiment. Thanks for pointing this out.

sunyatasattva commented on src/BlockTypes/ProductAddToCartButton.php:
```diff
@@ -0,0 +1,262 @@
+	}
+
+	/**
+	 * Callback function that modifies the HTML content of a "core/button" block.
+	 * The function checks if the block is a WooCommerce variation, fetches the product information, modifies the HTML structure of the block,
+	 * and applies the filters to the block content before returning it.
```
```suggestion
	 * Callback function that modifies the HTML content of a "core/button" block.
	 *
	 * The function checks if the block is a WooCommerce variation,
	 * fetches the product information, modifies the HTML structure of the block,
	 * and applies the filters to the block content before returning it.
```

sunyatasattva commented on src/BlockTypes/ProductAddToCartButton.php:
```diff
@@ -0,0 +1,262 @@
+			$product = wc_get_product( $post_id );
+
+			if ( $product ) {
+				$styles_and_classes = $this->extract_style_and_class_from_block_content( $block_content );
+				$div_class          = $styles_and_classes['div_class'];
+				$div_style          = $styles_and_classes['div_style'];
+				$anchor_class       = $styles_and_classes['anchor_class'];
+				$anchor_style       = $styles_and_classes['anchor_style'];
```
You could destructure this like so:

```suggestion
				[
				  'div_class' => $div_class,
				  'div_style' => $div_style,
				  'anchor_class' => $anchor_class,
				  'ancor_style' => $anchor_style
				] = $this->extract_style_and_class_from_block_content( $block_content );
```

imanish003 replied:
Oh, My reason behind not using destructring was that array destructing is available only in PHP version 7.1 and above.
But looks like we have updated Requires PHP from 7.0 to 7.2 in [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/8154/files) 2 weeks ago, so now it's safe to use "array unpacking" or "destructuring assignment".

Thanks for pointing this out 🙏🏻

sunyatasattva commented on src/BlockTypes/ProductAddToCartButton.php:
```diff
@@ -0,0 +1,262 @@
+
+
+	/**
+	 * Return inner html of anchor block. Some examples of inner html of `a` tag:
+	 * - <strong><em><mark><code>Add to cart</code></mark></em></strong>
+	 * - <strong>Add</strong> <strong><em>to</em></strong> <em>cart</em>
```
```suggestion
	 * Return inner html of anchor block.
	 * 
	 * Due to the rich text capabilities of the Button block, the markup can look
	 * quite convoluted. Here are some examples of the inner HTML of `a` tag:
	 * ```html
	 * - <strong><em><mark><code>Add to cart</code></mark></em></strong>
	 * - <strong>Add</strong> <strong><em>to</em></strong> <em>cart</em>
	 *```
```

imanish003 replied:
Thanks for making this comment more clear 🙌🏻

sunyatasattva reviewed (changes requested):
Excellent work, Manish! I've got a few comments here and a few more in the P2 discussion. Thanks for the excellently clear code with explicit comments!

imanish003 committed:
use useEffect instead of setTimeout to hide link toolbar

imanish003 committed:
remove feature flag, icon and other minor improvements

imanish003 committed:
Merge branch 'trunk' into try/using-buttons-for-add-to-cart-2

imanish003 committed:
Add text styling for simple products only

Due to the rich text capabilities of the Button block, the markup can look quite convoluted. Here are some examples of the inner HTML of `a` tag:
```html
- <strong><em><mark><code>Add to cart</code></mark></em></strong>
- <strong>Add</strong> <strong><em>to</em></strong> <em>cart</em>
```

This way merchant can style the text as they wish for the text shown for simple products.

For other product types, $product->add_to_cart_text() is a method that retrieves the text that should be displayed on the "Add to Cart" button.
For example, If the product is a variable product, text will be "Select options". If the product is out of stock, the text will be "Out of stock".

If the product is not a simple product, than we don't show any text styling because text styling is not supported for non-simple products because there is no way to style the text for non-simple products.

imanish003 committed:
Fix a typo

imanish003 committed:
Merge branch 'trunk' into try/using-buttons-for-add-to-cart-2

imanish003 committed:
Fix: Sometimes link toolbar button is not hiding

Sometimes editor.BlockListBlock filter is not called when the toolbar menu re-renders therefore we also need to hide the link toolbar button onClick event.

More info: https://github.com/woocommerce/woocommerce-blocks/pull/8267#discussion_r1088883081

imanish003 committed:
Add suffix CORE_NAME & VARIATION_NAME with names

More info: https://github.com/woocommerce/woocommerce-blocks/pull/8267#pullrequestreview-1271124110

imanish003 committed:
Using CSS to hide link toolbar button

Adding style to <head> when variation is selected & removing style when variation unmounts.
